File: src/data/sequences.py
from __future__ import annotations

from typing import Iterable, List, Sequence, Tuple, Union

from Bio import SeqIO


SequenceRecord = Tuple[str, str]


def load_fasta_sequences(
    file_path: str,
    *,
    remove_padding: bool = True,
    load_identifiers: bool = False,
) -> Union[List[str], List[SequenceRecord]]:
    """Load sequences from a FASTA file.

    Args:
        file_path: Path to FASTA file.
        remove_padding: Remove '-' alignment gaps when True.
        load_identifiers: Return (id, sequence) tuples instead of sequences.
    """
    records = SeqIO.parse(file_path, "fasta")
    return read_sequences(
        records,
        remove_padding=remove_padding,
        load_identifiers=load_identifiers,
    )


def separate_identifiers(train: Sequence[SequenceRecord]) -> Tuple[List[str], List[str]]:
    identifiers = [record_id for record_id, _ in train]
    sequences = [sequence for _, sequence in train]
    return identifiers, sequences


def load_fasta_msa(file_path: str, *, load_identifiers: bool = False):
    return load_fasta_sequences(
        file_path,
        remove_padding=False,
        load_identifiers=load_identifiers,
    )


def read_sequences(
    records: Iterable,
    *,
    remove_padding: bool = True,
    load_identifiers: bool = False,
) -> Union[List[str], List[SequenceRecord]]:
    sequences: Union[List[str], List[SequenceRecord]] = []
    for record in records:
        seq = str(record.seq)
        if remove_padding:
            seq = seq.replace("-", "")
        if load_identifiers:
            sequences.append((str(record.id), seq))
        else:
            sequences.append(seq)
    return sequences

File: src/data/legacy/julia/sequences.jl
using FASTX

#=
for record in reader
    record |> identifier
    record |> sequence
end
=#

function load_fasta_sequences(file_path::String; remove_padding::Bool=true, load_identifiers::Bool=false)
    reader = FASTAReader(open(file_path))
    sequences = read_sequences(reader; remove_padding, load_identifiers)
    close(reader)
    return sequences
end

function separate_identifiers(train)
    train_identifiers = getfield.(train, 1)
    train = getfield.(train, 2)
    return train_identifiers, train
end

function load_fasta_msa(file_path::String; load_identifiers::Bool=false)
    reader = FASTAReader(open(file_path))
    msa = read_sequences(reader; remove_padding=false, load_identifiers)
    close(reader)
    reduce(hcat, collect.(msa)) |> permutedims
end

function read_sequences(reader::FASTX.FASTA.Reader; remove_padding::Bool=true, load_identifiers::Bool=false)
    sequences = load_identifiers ? Vector{Tuple{String,String}}([]) : Vector{String}([])
    for record in reader
        seq = record |> FASTX.sequence |> string
        if remove_padding
            seq = replace(seq, "-" => "")
        end
        if load_identifiers
            id = record |> identifier |> string
        end
        load_identifiers ? append!(sequences, [(id, seq)]) : append!(sequences, [seq])
    end
    return sequences
end
